Lf 2986 sensor position changes when there are multiple sensors at the same location and tooltip is opened

javiscript13 opened the pull request:
**Description**

The issue is that the map and markers are created with the zoom of the location of the farm (a point).  But, there are some points (water valves, sensors and gates) that could be in a place where there is a different zoom.  So when the zoom of a group of points is greater than the zoom of the farm the marker will no longer appear, and when the zoom of a group of points is smaller than the zoof of the farm, the click on the marker will not show the popup.

For fixing it, the maxZoom of the map will be calculated based on all points (areas and lines not affected), so the maxZoom of the map is the smaller of the group of points.  If the user adds a new point the maxZoom will be recalculated.  To calculate the maxZoom for a given Point google map api has an endpoint where you check the maxZooms, so, when a new point is added only the maxZoom of the new points is queried and compared to the previous value.

This code fixes that issue on the main map, and on the task map (where the issue was similar but more manageable for the user).   The map for creating farm was reviewed but didn't have the issue and was not affected.

Jira link: https://lite-farm.atlassian.net/browse/LF-2986  and https://lite-farm.atlassian.net/browse/LF-3175

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Passes test case
- [x] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

SayakaOno reviewed (commented):
@javiscript13 
I'm still reviewing, but do you mind explaining the difference between "map" and "maps"? (`getMaxZoom`'s parameters)
Also, could we create a constant for the number "18" and add a comment on why it is 18?

javiscript13 commented:
> @javiscript13 I'm still reviewing, but do you mind explaining the difference between "map" and "maps"? (`getMaxZoom`'s parameters) Also, could we create a constant for the number "18" and add a comment on why it is 18?

Hey  @SayakaOno , thank you for reviewing this pull request.  

the value 18 used for fallback has been committed as a constant.

About "map" and "maps":

Both the map and maps objects are returned by the GoogleMap component on the api loaded event "onGoogleApiLoaded" .  (https://github.com/google-map-react/google-map-react/blob/master/API.md )

"Map" is the actual map that has been loaded.  It holds all the map properties like zoom, center, bounds, etc.
"Maps" is the google-maps-api object, so you can access all the api functions through it. So, for example, you can use it to draw things on the map, or accessing the "getMaxZoomAtLatLng" which is an API function.

Before the changes on this PR, getMaxZoom only needed "maps" to retrieve the maxZoom of the farm location using "getMaxZoomAtLatLng".  It would then set the maxZoom of the mapCacheReducer redux-state, and this value would be used on the click-event of the marker to check if it should display the popup. The "map" would have a default maxZoom value of 80 (no map location has that maxZoom as the current max is 21).

On this PR I'm adding the map parameter to getMaxZoom function to be able to set the map.maxZoom to the desired value by farm. So, the getMaxZoom function will retrieve all maxPoints and use the smaller one to execute this:  map.setOptions({ maxZoom: minNumber })  (it is also used on a catch statement). The map.setOptions(...) statement will prevent that the map zoom more than the set value in any place of the map.

Please let me know any other needed detail.

SayakaOno reviewed (commented):
@javiscript13 thank you for providing the detailed information! I have a better understanding now.

Great job on figuring things out, your code looks great overall.
I see an issue that points don't cluster at first where they should. Does it happen to you?

https://github.com/LiteFarmOrg/LiteFarm/assets/33141219/a8582b94-5218-4e04-a806-440053ffa961

javiscript13 commented:
> *

Hey @Duncan-Brain, thank you for reviewing this PR!

So, I have been trying to reproduce this bug, but after a while and reviewing again the video, I realized is not a bug but the implemente feature:   

You have a couple points very near this place: 

![image](https://github.com/LiteFarmOrg/LiteFarm/assets/4588278/e03b7da8-01c9-4e24-8b3d-aa890046be0e)

and as you can see on the image, the maxZoom for that place is 5, (could be 6 since there are places where it is that other value. You can check it on: https://developers.google.com/maps/documentation/javascript/examples/maxzoom-simple).  So, what is happening is that the maxZoom will take the minimum max zoom of your farm, and as this value is like 5 for those 2 points, the maxZoom will have that very small value.  If you delete those 2 points, and keep the one on the ocean it will have a maxZoom between 9 and 10 (which are the maxValues that most of the ocean has).

About the useEffect, after thinking for a while I think it is best to change as you suggested. It has been added in a commit.

javiscript13 commented on packages/webapp/src/components/Map/MarkerCluster/index.jsx:
```diff
@@ -14,12 +14,13 @@ const CreateMarkerCluster = (
     map,
     markers,
-    algorithm: new SuperClusterAlgorithm({ maxZoom: 20, radius: 45.5 }),
+    algorithm: new SuperClusterAlgorithm({ maxZoom: maxZoom, radius: 45.5 }),
```
Hey @SayakaOno , not quite correct.  The map actual maxZoom is set by the setOptions, this maxZoom is being sent to the SuperClusterAlgorithm that creates the markers.  Funny things happens to the markers when they are not in sync.  But as a map can have no markers, the actual maxZoom for the map has to be set by the map options.

SayakaOno replied:
I have a much better understanding now! Thank you :)

This finally made sense! Yay!!
https://github.com/mapbox/supercluster#options

javiscript13 commented on packages/webapp/src/containers/Map/useMaxZoom.js:
```diff
@@ -1,22 +1,76 @@
+          !retrievedPoints.some(
+            (item) => item.lat === element.point.lat && item.lng === element.point.lng,
+          )
+        ) {
```
great simplification!!

javiscript13 commented on packages/webapp/src/containers/Map/useMaxZoom.js:
```diff
@@ -1,22 +1,76 @@
       });
+      const promises = pointsToQuery.map(function (point) {
+        return new Promise(function (resolve, reject) {
+          mapService?.getMaxZoomAtLatLng(point, function (result) {
```
If there is no mapService, the case will be handled by the catch on the promises. I added a fix on that catch because it was setting the redux-store maxZoom but not the map maxZoom, now fixed, if there is a problem with mapService the DEFAULT_MAX_ZOOM is set inside the catch sentence.

javiscript13 replied:
on error, if available the maxZoom will be retrieved from the stored cache of MaxZooms, if not, then DEFAULT_MAX_ZOOM will be used.

javiscript13 commented:
Hey @SayakaOno and @Duncan-Brain:

From the last time you reviewed this, and in addition to the changes due to your comments, I made a couple changes that I consider "big".

1. The one that requires more attention is that there were a useEffect with "sensors" as dependency array.  I changed that to maxZoom.  The original useEffect updated the sensors on the map, but ocassionally the markers didn't reflect the changes because maxZoom wasn't updated at the moment recreate markers.  So I changed it to update when the maxZoom was available, which also works for the sensors since maxZoom is updated every time a sensor is created or deleted. 

2.  I changed the way I was handling points maxZoom cache, since I wasn't considering that a point could be deleted I was storing only calculations, but now I'm storing maxZoom per point, so if a point is deleted the maxZoom will be recalculated using cache with only the current points.

Thanks!

javiscript13 commented:
> @javiscript13 thank you for providing the detailed information! I have a better understanding now.
> 
> Great job on figuring things out, your code looks great overall. I see an issue that points don't cluster at first where they should. Does it happen to you?
> 
>  Screen.Recording.2023-05-16.at.2.11.40.PM.mov

Hey @SayakaOno I wasn't able to recreate this bug exactly as your video, but it happened sometimes that the markers were not recreated correctly, as mentioned on my comment above, I made some changes on a useEffect with dependency array sensors (now maxZoom).  Please let me know if that fixes this.

Thx.

SayakaOno reviewed (approved):
> > @javiscript13 thank you for providing the detailed information! I have a better understanding now.
> > Great job on figuring things out, your code looks great overall. I see an issue that points don't cluster at first where they should. Does it happen to you?
> > Screen.Recording.2023-05-16.at.2.11.40.PM.mov
> 
> Hey @SayakaOno I wasn't able to recreate this bug exactly as your video, but it happened sometimes that the markers were not recreated correctly, as mentioned on my comment above, I made some changes on a useEffect with dependency array sensors (now maxZoom). Please let me know if that fixes this.
> 
> Thx.

@javiscript13 
I no longer see this issue, and your additional changes look good to me.
Great job!

Duncan-Brain reviewed (approved):
Hi @javiscript13 , So chatted with @kcussen and he agreed that while it could possibly be an issue for some farms near low imagery areas we can handle the new issue in a separate ticket please see (https://lite-farm.atlassian.net/browse/LF-3304) .

I thought it would be good to add a comment with some extra info on the jira ticket that you provided so I uploaded that csv as well.

Thanks for your patience here!
